### PR TITLE
kmsv2: KDF based nonce extension

### LINKED ISSUE #118831: kmsv2: fix limitations caused by counter based nonce
xref: https://hackmd.io/@enj/SyiXCABZn

### PR DESCRIPTION
**Pull request by enj**
Fixes #118831

/kind feature
/kind api-change
/sig auth
/triage accepted
/milestone v1.28

```release-note
Changed how KMS v2 encryption at rest can generate data encryption keys.
When you enable the `KMSv2KDF` feature gate (off by default), KMS v2 uses a key derivation function to generate single use data encryption keys from a secret seed combined with some random data.  This eliminates the need for a counter based nonce while avoiding nonce collision concerns associated with AES-GCM's 12 byte nonce.
```

```docs
TODO link PR to update docs
```

### CODE REVIEW

**Comment by zelenushechka** [comment]
Hey, I'm Anhelina from the Bug Triage Release team 1.28 👋. I wanted to let you know that Code Freeze is coming up on [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496), Please make sure that this PR is submitted before the deadline. We are currently in Week 7, and the Code Freeze will be in effect by Week 9.
Do let me know if I can be of any help :)

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go [hunk 1]
```diff
@@ -177,25 +199,153 @@ func generateKey(length int) (key []byte, err error) {
...
+// NewKDFExtendedNonceGCMTransformerFromSeedUnsafe is the same as NewGCMTransformer but trades storage,
+// memory and CPU to work around the limitations of AES-GCM's 12 byte nonce size.  It is marked
+// as unsafe because it assumes that the input seed is a cryptographically strong key that is at least
+// 32 bytes in length.  Callers that do need to supply a specific seed must use
+// NewKDFExtendedNonceGCMTransformerWithUniqueSeed instead to guarantee that a secure seed is used.
+// Unlike NewGCMTransformer, this function is immune to the birthday attack because a new key is generated
+// per encryption via a key derivation function: KDF(seed, random_bytes) -> key.  The derived key is
+// only used once as an AES-GCM key with a random 12 byte nonce.  This avoids any concerns around
+// cryptographic wear out (by either number of encryptions or the amount of data being encrypted).
+// No specific rotation schedule is required for the seed.  NewKDFExtendedNonceGCMTransformerWithUniqueSeed
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:211]: @mikedanese do you agree with the statements around seed rotation:

> No specific rotation schedule is required for the seed.

> reusing the seed returned from this function is safe to do over time and across process restarts.

I am still working on confirming if MS crypto folks have an opinion on how often (if ever) the seed needs to be rotated (assuming the KEK is not being rotated / rotated very rarely - KEK rotation synthetically causes the seed to rotate).

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 2]
```diff
@@ -346,32 +347,38 @@ func (h *kmsv2PluginProbe) rotateDEKOnKeyIDChange(ctx context.Context, statusKey
...
-
-	// state is valid and status keyID is unchanged from when we generated this DEK so there is no need to rotate it
+	// allow writes for only up to kmsv2PluginWriteDEKorSeedMaxTTL from now when there are errors
+	// we start the timer before we make the network call because kmsv2PluginWriteDEKorSeedMaxTTL is meant to be the upper bound
+	expirationTimestamp := envelopekmsv2.NowFunc().Add(kmsv2PluginWriteDEKorSeedMaxTTL)
+
+	// dynamically check if we want to use KDF seed to derive DEKs or just a single DEK
+	// this gate can only change during tests, but the check is cheap enough to always make
+	// this allows us to easily exercise both modes without restarting the API server
+	// TODO integration test that this dynamically takes effect
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:357]: Not a blocker for this PR.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 3]
```diff
@@ -42,6 +43,8 @@ import (
 	"k8s.io/utils/clock"
 )
 
+// TODO integration test with old AES GCM data recorded and new KDF data recorded
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:46]: Not a blocker for this PR.

**Comment by enj** [comment]
@deads2k @mikedanese I will address any comments after I get back from PTO, but this is ready.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go [hunk 4]
```diff
@@ -177,25 +199,153 @@ func generateKey(length int) (key []byte, err error) {
...
+// and this function encrypt and decrypt data in a compatible way (so the output of one can be used
+// as the input to the other, and vice versa).
+func NewKDFExtendedNonceGCMTransformerFromSeedUnsafe(seed []byte) (value.Transformer, error) {
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:214]: Oh I see this being used in the kmsv2 envelop.go

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go [hunk 5]
```diff
@@ -177,25 +199,153 @@ func generateKey(length int) (key []byte, err error) {
...
+
 func (t *gcm) TransformFromStorage(ctx context.Context, data []byte, dataCtx value.Context) ([]byte, bool, error) {
 	nonceSize := t.aead.NonceSize()
-	if len(data) < nonceSize {
+	infoLen := len(t.info)
+	if len(data) < infoLen+nonceSize {
 		return nil, false, errors.New("the stored data was shorter than the required size")
 	}
-	result, err := t.aead.Open(nil, data[:nonceSize], data[nonceSize:], dataCtx.AuthenticatedData())
+	result, err := t.aead.Open(nil, data[infoLen:infoLen+nonceSize], data[infoLen+nonceSize:], dataCtx.AuthenticatedData())
```
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:332]: Note: I don't think this is compatible with FIPS. For A.5.2 of "Key/IV Pair Uniqueness Requirements from SP 800-38D", I believe we need to generate random IVs internal to the FIPS module.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go [hunk 6]
```diff
@@ -23,20 +23,38 @@ import (
...
+	"golang.org/x/crypto/hkdf"
+
 	"k8s.io/apiserver/pkg/storage/value"
 	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
 )
 
+// commonSize is the length of various security sensitive byte slices such as encryption keys.
+// Do not change this value.  It would be a backward incompatible change.
+const commonSize = 32
```
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:43]: nit: Break this constant up into keySize, seedSize, kdfNonceSize, etc... even if they are all 32 so that they are easier to make sense of in context.

Right now reading code like:
```
info := data[:commonSize]
```
Has a decent amount of cognitive overhead.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go [hunk 7]
```diff
@@ -177,25 +199,153 @@ func generateKey(length int) (key []byte, err error) {
...
+// NewKDFExtendedNonceGCMTransformerWithUniqueSeed is the same as NewKDFExtendedNonceGCMTransformerFromSeedUnsafe
+// but it handles the seed generation for the caller.  Whenever a new seed is needed (for example,
+// during key rotation), this function should be used over the caller generating a new seed.  This
+// function is considered safe because there is no input that a caller could make a mistake with,
+// and the output transformer has the properties of AES-GCM without the nonce size limitations.
+// A new random seed is generated and returned on every invocation of this function.  If the seed is
+// stored and retrieved at a later point, it can be passed to NewKDFExtendedNonceGCMTransformerFromSeedUnsafe
+// to construct a transformer capable of decrypting values encrypted by this transformer;
+// reusing the seed returned from this function is safe to do over time and across process restarts.
+func NewKDFExtendedNonceGCMTransformerWithUniqueSeed() (value.Transformer, []byte, error) {
```
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:233]: Why do we need a function to internally generate the key? Can we move this to match the NewCBCTransformer API/usage now?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go [hunk 8]
```diff
@@ -23,20 +23,38 @@ import (
...
 	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
 )
 
+// commonSize is the length of various security sensitive byte slices such as encryption keys.
+// Do not change this value.  It would be a backward incompatible change.
+const commonSize = 32
+
+// cacheTTL is the TTL of KDF cache entries.  We assume that the value.Context.AuthenticatedData
+// for every call is the etcd storage path of the associated resource, and use that as the primary
```
[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:46]: The cache key is `{AD, info}` so we rely on the `AD` being one per individual resource.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go (hunk 5 above)
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:332]: I do not follow - how is this breaking FIPS rules exactly?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go (hunk 7 above)
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:233]: Sure I can collapse onto the one function that takes the key as input.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go [hunk 9]
```diff
@@ -177,25 +199,153 @@ func generateKey(length int) (key []byte, err error) {
 	return key, nil
 }
 
+// NewKDFExtendedNonceGCMTransformerFromSeedUnsafe is the same as NewGCMTransformer but trades storage,
+// memory and CPU to work around the limitations of AES-GCM's 12 byte nonce size.  It is marked
+// as unsafe because it assumes that the input seed is a cryptographically strong key that is at least
+// 32 bytes in length.  Callers that do need to supply a specific seed must use
```
[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:205]: That is fair, will update.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go [hunk 10]
```diff
@@ -23,20 +23,38 @@ import (
...
+	"k8s.io/utils/clock"
 )
 
+// commonSize is the length of various security sensitive byte slices such as encryption keys.
+// Do not change this value.  It would be a backward incompatible change.
+const commonSize = 32
+
+// cacheTTL is the TTL of KDF cache entries.  We assume that the value.Context.AuthenticatedData
+// for every call is the etcd storage path of the associated resource, and use that as the primary
+// cache key (with a secondary check that confirms that the gcm.info matches).  Thus if a client
```
[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:47]: It is a mix of code re-use / efficiency / caching logic.

[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:47]: In particular, I wanted to avoid the allocations associated with appending the `info` and ciphertext separately.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go (hunk 5 above)
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:332]: The `info` is still generated within this package and is hidden from all external callers (only private functions can set it).  That seems fine to me?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:47]: Okay I moved the `info` out which led to `+2.62%` on `B/op` but no other changes.  Seems like an okay tradeoff.

```
            │ /tmp/old_stor │            /tmp/new_stor            │
            │    sec/op     │    sec/op     vs base               │
KMSv2KDF-10     9.094 ± 13%   10.009 ± 13%  ~ (p=0.853 n=10)

            │ /tmp/old_stor │            /tmp/new_stor            │
            │     B/op      │     B/op      vs base               │
KMSv2KDF-10    17.90Gi ± 0%   18.37Gi ± 0%  +2.62% (p=0.000 n=10)

            │ /tmp/old_stor │            /tmp/new_stor            │
            │   allocs/op   │  allocs/op    vs base               │
KMSv2KDF-10     1.159M ± 1%   1.168M ± 1%   ~ (p=0.218 n=10)
```

[the remark this answers is not in the document]
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:47]: Ya... The "crytop/cipher".AEAD interface is pretty close to ideal for this situation: https://pkg.go.dev/crypto/cipher#AEAD. Might be worth exploring if that's a better API to build on (e.g. add Overhead() and dst buffers to value.Transform).

[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:47]: @mikedanese okay I also implemented an approach based on `cipher.AEAD` which is slightly faster than the other approach but needs a different data layout.  Which do you prefer?  I will have to update some tests for the `cipher.AEAD` approach but I will wait until we decide if that is the way we want to go.

[transformerWithInfo - data stored as info|nonce|ciphertext - uses a value.Transformer wrapper](https://github.com/enj/kubernetes/blob/06b98a816d53233d1cf514929f80367721a57ec1/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes_extended_nonce.go)

[aeadWithInfo - data stored as nonce|info|ciphertext - uses a cipher.AEAD wrapper](https://github.com/enj/kubernetes/blob/66bb520c40874230675af6aae79ab6428e389794/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes_extended_nonce.go)

[diff between the two approaches](https://github.com/kubernetes/kubernetes/compare/06b98a816d53233d1cf514929f80367721a57ec1...66bb520c40874230675af6aae79ab6428e389794)

[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:47]: BTW I am leaning towards `transformerWithInfo` since it is most similar to other code in k/k and the implementation is easier to follow.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go (hunk 1 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:211]: I would rephrase this as, "Speaking on the cryptographic safety, the limit on the number of operations that can be preformed with a single key with randomly generated nonces is not practically reachable. Thus, the scheme does not impose any specific requirements on the key rotation schedule."

Then say an organization standards or compliance policies around rotation can be implemented externally by rotating the kek.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go [hunk 11]
```diff
@@ -54,12 +72,16 @@ type gcm struct {
 // rotation. Future work should include investigation of AES-GCM-SIV as an alternative to
 // random nonces.
```
[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:73]: I will leave this for now and will open a future PR to drop this.

File: test/integration/controlplane/transformation/kmsv2_transformation_test.go [hunk 12]
```diff
@@ -642,3 +837,217 @@ resources:
...
+	}
+	b.Cleanup(test.cleanUp)
+
+	client := kubernetes.NewForConfigOrDie(test.kubeAPIServer.ClientConfig)
+
+	if restOptionsGetter == nil {
+		b.Fatal("not REST options found for secrets")
+	}
+
+	secretStorage, err := secretstore.NewREST(restOptionsGetter)
```
[the remark this answers is not in the document]
Comment (enj) [reply test/integration/controlplane/transformation/kmsv2_transformation_test.go:884]: I came up with an approach to construct the secret store without needing to mess with the etcd options code.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go (hunk 1 above)
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go]: > I think the recommendation here is to rotate DEK/seed every month.

My reading of the MS crypto folks recommendations was that 'we could rotate the seed every month just to make sure it all works.'  I will circle back with them on this.  If need be we can update the docs next release (or update the code to do automatic monthly seed rotation).

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/v2/api.proto [hunk 13]
```diff
@@ -29,8 +29,13 @@ message EncryptedObject {
   string keyID = 2;
 
   // EncryptedDEK is the encrypted DEK.
+  // Mutually exclusive with encryptedSeed.
   bytes encryptedDEK = 3;
```
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/v2/api.proto:33]: can we add a DEK type instead of adding a seed field?

```
enum Type {
   AES_GCM = 1;
   XAES_GCM_HKDF_SHA256 = 2;
}
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/cache.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+	cache.AllowExpiredGet = true
+	return &simpleCache{
+		cache: cache,
+		ttl:   ttl,
+	}
+}
+
+// given a key, return the transformer, or nil if it does not exist in the cache
+func (c *simpleCache) get(info []byte, dataCtx value.Context) *transformerWithInfo {
+	val, ok := c.cache.Get(keyFunc(dataCtx))
```
[the remark this answers is not in the document]
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/cache.go:50]: Resolved on slack. AD is used to free cache entries conveniently when objects get overwritten.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/v2/api.proto (hunk 13 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/v2/api.proto:33]: Agree on an enum to distinguish the meaning of the field rather than a separate field containing encrypted

If we want to rename the `encryptedDEK` field while keeping the same proto tag, that's fine and wire-compatible. I like the `encrypted` prefix, and I like including `DEK` somewhere in the name since the contents of this are used as the DEK or to derive the DEK. Bad ideas that will maybe spawn good ideas:
* `encryptedDEKInput`
* `encryptedDEKSource`
* `encryptedDEK___`

Name the enum field something that matches the `encryptedDEK` field name, e.g. `encryptedDEKType`

For the enum values, we could include the details of what the contents are (e.g., for the KDF use, it's a seed):

```
AES_GCM_...
HKDF_SHA256_XNONCE_AES_GCM_SEED
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go [hunk 15]
```diff
@@ -164,8 +131,8 @@ func die(msg string) {
 	klog.FatalDepth(1, msg)
 }
 
-// generateKey generates a random key using system randomness.
-func generateKey(length int) (key []byte, err error) {
+// GenerateKey generates a random key using system randomness.
+func GenerateKey(length int) (key []byte, err error) {
```
[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:135]: Can do in a follow up - internally this sets metrics right now so would need to split this up to have a `GetRandomBytes` func.

**Comment by liggitt** [comment]
lgtm/approve for the proto API update (@mikedanese can weigh in on enum names, or we can tweak the text names in the future without changing the wire serialization)

I didn't review the implementation, will leave that review to mike

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/v2/api.proto [hunk 16]
```diff
@@ -28,9 +28,24 @@ message EncryptedObject {
...
+  bytes encryptedDEKSource = 3;
 
   // Annotations is additional metadata that was provided by the KMS plugin.
   map<string, bytes> annotations = 4;
+
+  // encryptedDEKSourceType defines the process of using the plaintext of encryptedDEKSource to determine the DEK.
+  EncryptedDEKSourceType encryptedDEKSourceType = 5;
+}
+
+enum EncryptedDEKSourceType {
```
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/v2/api.proto:42]: Minor: I might drop the "Encrypted" prefix. This describes a property of the unencyrpted DEKSource, it shows up in code around the unencrypted key, etc... encryptedDEKSourceType sounds fine as the field name, I'd just drop it from the enum type name.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 17]
```diff
@@ -214,39 +221,43 @@ func (t *envelopeTransformer) TransformToStorage(ctx context.Context, data []byt
...
-	return t.doEncode(encObject)
+	return t.doEncode(&encObjectCopy)
 }
 
 // addTransformerForDecryption inserts a new transformer to the Envelope cache of DEKs for future reads.
-func (t *envelopeTransformer) addTransformerForDecryption(cacheKey []byte, key []byte) (value.Read, error) {
-	block, err := aes.NewCipher(key)
-	if err != nil {
-		return nil, err
+func (t *envelopeTransformer) addTransformerForDecryption(cacheKey []byte, key []byte, useSeed bool) (value.Read, error) {
```
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:234]: We should pass in the EncryptedDEKSourceType and switch vs useSeed. That would be the factoring e.g. once we support three types.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 18]
```diff
@@ -258,18 +269,31 @@ func (t *envelopeTransformer) doDecode(originalData []byte) (*kmstypes.Encrypted
...
 	return o, nil
 }
 
-// GenerateTransformer generates a new transformer and encrypts the DEK using the envelope service.
-// It returns the transformer, the encrypted DEK, cache key and error.
-func GenerateTransformer(ctx context.Context, uid string, envelopeService kmsservice.Service) (value.Transformer, *kmsservice.EncryptResponse, []byte, error) {
-	transformer, newKey, err := aestransformer.NewGCMTransformerWithUniqueKeyUnsafe()
+// GenerateTransformer generates a new transformer and encrypts the DEK/seed using the envelope service.
+// It returns the transformer, the encrypted DEK/seed, cache key and error.
+func GenerateTransformer(ctx context.Context, uid string, envelopeService kmsservice.Service, useSeed bool) (value.Transformer, *kmstypes.EncryptedObject, []byte, error) {
```
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:281]: Same re useSeed.

**Comment by sftim** [comment]
Changelog suggestion:
```markdown
Changed KMS API encryption can generate data encryption keys.
When you enable the `KMSv2KDF` feature gate (off by default), KMS v2 uses a key derivation function to generate single use data encryption keys from a secret seed combined with some random data.  This eliminates the need for a counter based nonce while avoiding nonce collision concerns associated with AES-GCM's 12 byte nonce.
```

BTW, is there a docs PR to cover this change?
